core/jvm/src/main/scala/fs2/interop/flow/package.scala

```diff
@@ -0,0 +1,253 @@
+  *
+  * @example {{{
+  * scala> import cats.effect.{IO, Resource}
+  * scala> import fs2.Stream
+  * scala> import fs2.interop.flow._
+  * scala> import java.util.concurrent.Flow.Publisher
+  * scala>
+  * scala> val upstream: Stream[IO, Int] = Stream(1, 2, 3).covary[IO]
+  * scala> val publisher: Resource[IO, (Publisher[Int], Deferred[F, Unit])] = upstream.toPublisher
+  * scala> val downstream: Stream[IO, Int] = Stream.resource(publisher).flatMap {
+  *   case (publisher, completedSignal) =>
+  *     publisher.toStream[IO](bufferSize = 16)
+  * }
+  * scala>
+  * scala> import cats.effect.unsafe.implicits.global
+  * scala> downstream.compile.toVector.unsafeRunSync()
+  * res0: Vector[Int] = Vector(1, 2, 3)
+  * }}}
+  *
+  * @see [[java.util.concurrent.Flow]]
+  */
+package object flow {
+
+  /** Creates a [[Stream]] from a function exposing a [[Subscriber]].
+    *
+    * This function is useful when you actually need to provide a subscriber to a third-party.
+    * The subscription process is uncancelable.
+    *
+    * @example {{{
+    * scala> import cats.effect.IO
+    * scala> import fs2.Stream
+    * scala> import fs2.interop.flow.fromOnSubscribe
+    * scala> import java.util.concurrent.Flow.{Publisher, Subscriber}
+    * scala>
+    * scala> // This internally calls somePublisher.subscribe(subscriber)
+    * scala> def thirdPartyLibrary[A](subscriber: Subscriber[A]): Unit = ???
+    * scala>
+    * scala> def consumeThirdPartyLibrary[A]: Stream[IO, A] =
+    *   fromSubscriber[IO, A](bufferSize = 16) { subscriber =>
+    *     IO.println("Subscribing!") >>
+    *     IO.delay(thirdPartyLibrary(subscriber)) >>
+    *     IO.println("Subscribed!")
+    *   }
+    * }}}
+    *
+    * @note The publisher only receives a subscriber when the stream is run.
+    *
+    * @see [[fromPublisher]] for a simpler version that only requires a [[Publisher]].
+    *
+    * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+    *                   A high number can be useful if the publisher is triggering from IO,
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    * @param onSubscribe The effectual function to run during the subscription process,
+    *                    receives a [[Subscriber]] that should be used to subscribe to a [[Publisher]].
+    *                    The `subscribe` operation must be called exactly once.
+    */
+  def fromSubscriber[F[_], A](
+      bufferSize: Int
+  )(
+      subscribe: Subscriber[A] => F[Unit]
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    Stream
+      .eval(StreamSubscriber[F, A](bufferSize))
+      .flatMap { subscriber =>
+        subscriber.stream(
+          F.uncancelable(_ => subscribe(subscriber))
+        )
+      }
```

Comment:
Should we expose another version that gives access to `Poll` to users?

Reply:
Also, I think the name should be `fromSubscribe` as in "from a `subscribe` function". Arguably the name could also be `fromPublisher`, since `Subcriber[A] => Unit` is isomorphic to publisher.

Reply:
I don't remember why I was sure the `subscribe` bit should not be cancelable.

So, we can guarantee that on the other methods. Thus, we may also guarantee this on this one, or leave that to the discretion of users since this API is more low-level / error prone.

Reply:
> Also, I think the name should be fromSubscribe as in "from a subscribe function"

Well, TBH, that would be `fromOnSubscribe`, not this one. Because this function allows doing way more than just `subscribe`... even if anything else than just subscribe is unsafe; which is why I wasn't too keen to add this.

> Arguably the name could also be fromPublisher, since Subcriber[A] => Unit is isomorphic to publisher

Right, if we remove the other one I think is better to overload; otherwise I personally prefer the three different names.
Although, I wonder if it could lead to type inference issues.

Reply:
> Because this function allows doing way more than just `subscribe`

Not really :) the signature of this function is isomorphic to `Publisher`. Whatever you can do with one, you can do with the other.

> Although, I wonder if it could lead to type inference issues.

In this case it will be fine, because both the arity and types of the function are changing.

Reply:
> Not really :) the signature of this function is isomorphic to Publisher

Sure, which is why I do like `fromPublisher`, rather than `fromSubscribe`.